draft-ietf-quic-transport.md

```diff
@@ -861,67 +861,62 @@ different IP or port at either endpoint, due to NAT rebinding or mobility, as
-without connection IDs and long-form packets for connections that have
-incomplete cryptographic handshakes are associated with an existing
-connection using the tuple of source and destination IP addresses and ports.
-Other packets are associated with existing connections using connection ID
-the connection ID in the header; this might include connection IDs that were
-advertised using NEW_CONNECTION_ID ({{frame-new-connection-id}}).
-
-Clients SHOULD discard any packet that cannot be associated with an existing
-connection.  Discarded packets MAY be logged for diagnostic or security
-purposes.
-
-If a server receives a packet not associated with an existing connection, it
-executes the following steps, in order:
-
-1. The server MUST check if the packet uses a short form header, or is not
-long enough for the size required for the first packet of any QUIC version
-that the server supports. See {{packet-size}} for the definition of packet size
-and the minimum size of the first packet in this version of QUIC. If either
-condition is true, the packet cannot create a new connection. In this case, the
-server MUST either buffer the packet (see {{handshake-buffer}}), send a
-stateless reset ({{stateless-reset}}), or silently drop it.
-
-2. Otherwise, the server checks the version field in the long header. If the
-server does not support the chosen version, it MUST send a Version
-Negotiation packet as described in {{send-vn}}.
-
-3. If the server supports the version, the server operates in accordance with
-the specification of that version. For the version described in this
-specification, it checks if the packet is a correctly formatted Initial packet.
-If so, the server MUST proceed with the handshake ({{handshake}}).  This
-commits the server to the version that the client selected. If not an Initial
-packet, the server MUST either buffer the packet or silently drop it.
-
-### Handshake Buffering {#handshake-buffer}
-
-Due to packet reordering or loss, subsequent packets might arrive at the server
-prior to the Initial packet. As described above, servers MAY buffer these
-packets in anticipation of the Initial packet arriving later. This is
-especially useful for 0-RTT packets that routinely accompany Initial packets.
-
-Buffering ensures that data is not lost, which improves performance; conversely,
-discarding these packets could create false loss signals for the congestion
-controllers.  However, limiting the number and size of buffered packets might be
-needed to prevent exposure to denial of service.
-
-Servers MUST NOT send packets in response to these buffered packets until
-the initial packet arrives.
-
-In this version of QUIC, clients cannot send short form headers prior to a
-response from the server. Servers that do not support a version of QUIC that
-allows early short headers MUST either respond with a stateless reset or drop
-the packet silently.
+First, hosts try to associate the packet with an existing connection. If the
```

Comment:
Wouldn’t you first look at the packet type, in order to handle Initial packets statelessly?

Reply:
That's a good point.  It certainly makes sense for a server to push any stateless processing to the front.

I don't know if it makes sense to consider that as an optional amendment, or whether it isn't just easier to reorder the processing.  It depends on whether it is easier to say "a server might move any stateless processing of potential Initial packets to the start of this process", or whether it is easier to rearrange the order from the outset.

Reply:
"Initial" is a version-dependent concept, so you still check the version first.  If you understand the version, then you can check whether it's an Initial packet.  However, since that's for long headers, the version will always be explicitly present, so you don't have to look up the connection ID first.